### PR TITLE
[catloop] Move closer to Catnap arch

### PR DESCRIPTION
**Pull request by iyzhang**
This PR moves Catloop to the Catnap arch. It also moves all state machine management directly into the Socket data structure (the second commit), which eliminates all calls to prepare_* and commit and rollback.

### CODE REVIEW

File: src/rust/catloop/socket.rs [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+        };
+        self.local = Some(local);
+        Ok(())
+    }
+
+    /// Enables this socket to accept incoming connections.
+    pub fn listen(&mut self) -> Result<(), Fail> {
+        self.state.prepare(SocketOp::Listen)?;
+        self.state.commit();
+        // Nothing to do.
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:137]: Drop this comment.

File: src/rust/catloop/socket.rs [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+        loop {
+            // Grab next request from the control duplex pipe.
+            let new_qd: QDesc = match pop_magic_number(catmem.clone(), catmem_qd, &yielder).await {
+                // Received a valid magic number so create the new connection. This involves create the new duplex pipe
+                // and sending the port number to the remote.
+                Ok(true) => create_pipe(catmem.clone(), catmem_qd, &ipv4, new_port, &yielder).await?,
+                // Invalid request.
+                Ok(false) => continue,
+                // Some error.
+                Err(e) => return Err(e),
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:171]: If it fails, we should `abort()`.

File: src/rust/catloop/socket.rs [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+                },
+                // Invalid response.
+                Ok(false) => {
+                    // Clean up newly allocated duplex pipe.
+                    catmem.borrow().close(new_qd)?;
+                    continue;
+                },
+                // Some error.
+                Err(e) => {
+                    socket.borrow_mut().state.rollback();
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:191]: Should we also rollback socket allocation?

File: src/rust/catloop/socket.rs [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+        // Gets the port for the new connection from the server by sending a connection request repeatedly until a port
+        // comes back.
+        let result: Result<(QDesc, SocketAddrV4), Fail> = {
+            let new_port: u16 = get_port(catmem.clone(), ipv4, port, yielder).await?;
+
+            // Open underlying pipes.
+            let remote: SocketAddrV4 = SocketAddrV4::new(*ipv4, new_port);
+            let new_qd: QDesc = catmem.borrow().open_pipe(&format_pipe_str(ipv4, new_port))?;
+            // Send an ack to the server over the new pipe.
+            send_ack(catmem.clone(), new_qd, &yielder).await?;
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:222]: If we fail, we should `abort()`.

File: src/rust/catloop/socket.rs [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+        let catmem: Rc<RefCell<CatmemLibOS>> = socket.borrow().catmem.clone();
+
+        // Gets the port for the new connection from the server by sending a connection request repeatedly until a port
+        // comes back.
+        let result: Result<(QDesc, SocketAddrV4), Fail> = {
+            let new_port: u16 = get_port(catmem.clone(), ipv4, port, yielder).await?;
+
+            // Open underlying pipes.
+            let remote: SocketAddrV4 = SocketAddrV4::new(*ipv4, new_port);
+            let new_qd: QDesc = catmem.borrow().open_pipe(&format_pipe_str(ipv4, new_port))?;
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:220]: If we fail, we sould `abort()`

File: src/rust/catloop/socket.rs [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+            Ok((
+                self.do_generic_sync_control_path_call(coroutine, yielder)?
+                    .get_task_id()
+                    .into(),
+                true,
+            ))
+        }
+    }
+
+    pub fn do_close(socket: Rc<RefCell<Self>>) -> Result<(), Fail> {
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:289]: Missing documentation?

File: src/rust/catloop/socket.rs [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+            Some(qd) => self.catmem.borrow().push(qd, sga),
+            None => Err(Fail::new(libc::EINVAL, "pipe not connected")),
+        }
+    }
+
+    /// Pop from the underlying Catmem queue that is the transport for this socket.
+    pub fn pop(&self, size: Option<usize>) -> Result<QToken, Fail> {
+        match self.catmem_qd {
+            Some(qd) => self.catmem.borrow().pop(qd, size),
+            None => Err(Fail::new(libc::EINVAL, "pipe not connected")),
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:308]: Is this the same error that we return for the other LibOSes? Shouldn't it be `ENOTCONN`?

File: src/rust/catloop/socket.rs [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+            .catmem_qd
+            .expect("Must be a catmem queue in this state.");
+        loop {
+            // Grab next request from the control duplex pipe.
+            let new_qd: QDesc = match pop_magic_number(catmem.clone(), catmem_qd, &yielder).await {
+                // Received a valid magic number so create the new connection. This involves create the new duplex pipe
+                // and sending the port number to the remote.
+                Ok(true) => create_pipe(catmem.clone(), catmem_qd, &ipv4, new_port, &yielder).await?,
+                // Invalid request.
+                Ok(false) => continue,
```
Comment (ppenna) [inline src/rust/catloop/socket.rs:169]: If it fails, should we `rollback`?

**Review by ppenna — APPROVED** [review]

**Review by anandbonde — APPROVED** [review]
Looks good.

**Pull request merged**